crates/block-producer/src/mempool/dependency_graph.rs

```diff
@@ -0,0 +1,826 @@
+            self.parents.remove(&key);
+
+            let children = self.children.remove(&key).unwrap_or_default();
+
+            // Remove edges from children to this node.
+            for child in children {
+                if let Some(child) = self.parents.get_mut(&child) {
+                    child.remove(&key);
+                }
+            }
+        }
+
+        Ok(pruned)
+    }
+
+    /// Removes the set of nodes __and all descendents__ from the graph, returning all removed
+    /// values.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if any of the given nodes does not exist.
+    ///
+    /// This method is atomic.
+    pub fn purge_subgraphs(&mut self, keys: BTreeSet<K>) -> Result<Vec<V>, GraphError<K>> {
+        let missing_nodes = keys
+            .iter()
+            .filter(|key| !self.vertices.contains_key(key))
+            .cloned()
+            .collect::<BTreeSet<_>>();
+        if !missing_nodes.is_empty() {
+            return Err(GraphError::UnknownNodes(missing_nodes));
+        }
+
+        let mut visited = keys.clone();
+        let mut to_process = keys;
+        let mut removed = Vec::new();
+
+        while let Some(key) = to_process.pop_first() {
+            let value = self
+                .vertices
+                .remove(&key)
+                .expect("Node was checked in precondition and must therefore exist");
+            removed.push(value);
+
+            self.processed.remove(&key);
+            self.roots.remove(&key);
+
+            // Children must also be purged. Take care not to visit them twice which is
+            // possible since children can have multiple purged parents.
+            let unvisited_children = self.children.remove(&key).unwrap_or_default();
+            let unvisited_children = unvisited_children.difference(&visited).cloned();
+            to_process.extend(unvisited_children);
+
+            // Inform parents that this child no longer exists.
+            let parents = self.parents.remove(&key).unwrap_or_default();
+            for parent in parents {
+                if let Some(parent) = self.children.get_mut(&parent) {
+                    parent.remove(&key);
+                }
+            }
+        }
+
+        Ok(removed)
+    }
+
+    /// Adds the node to the `roots` list _IFF_ all of its parents are processed.
+    ///
+    /// # SAFETY
+    ///
+    /// This method assumes the node exists. Caller is responsible for ensuring this is true.
+    fn try_make_root(&mut self, key: K) {
+        debug_assert!(self.vertices.contains_key(&key), "Potential root must exist in the graph");
+
+        let parents_completed = self
+            .parents
+            .get(&key)
+            .into_iter()
+            .flatten()
+            .all(|parent| self.processed.contains(parent));
+
+        if parents_completed {
+            self.roots.insert(key);
+        }
+    }
+
+    /// Set of nodes that are ready for processing.
```

Comment:
nit: I would start this with an action verb - e.g., "Returns a set of nodes ... ".

Reply:
Thanks. As an aside I believe there should be a way of connecting `roots()` and `process_root` via the type system. Something similar to a guard type for locks, which lets you choose a root(s) from the set and call `.process()` on it.

Not super critical but might be nice if its short code.

Won't look into this now; but when I'm feeling like type-golf sometime I'll pick at it.